node/src/service.rs

```diff
@@ -474,34 +499,86 @@ where
-	sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
-	+ sp_api::Metadata<Block>
-	+ sp_session::SessionKeys<Block>
-	+ sp_api::ApiExt<Block>
-	+ sp_offchain::OffchainWorkerApi<Block>
-	+ sp_block_builder::BlockBuilder<Block>
-	+ cumulus_primitives_core::CollectCollationInfo<Block>
-	+ sp_consensus_aura::AuraApi<Block, <<AuraId as AppCrypto>::Pair as Pair>::Public>
-	+ pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>
-	+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>
-	+ cumulus_primitives_aura::AuraUnincludedSegmentApi<Block>
-{
+#[cfg(not(feature = "ismp"))]
+mod runtime_api_ext {
+	use pop_runtime_common::{AccountId, AuraId, Balance, Block, Nonce};
+	use sp_core::Pair;
+	use sp_runtime::app_crypto::AppCrypto;
+
+	// Trait alias refactored out from above to simplify repeated trait bounds
+	pub(crate) trait RuntimeApiExt<RuntimeApi>:
+		sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+		+ sp_api::Metadata<Block>
+		+ sp_session::SessionKeys<Block>
+		+ sp_api::ApiExt<Block>
+		+ sp_offchain::OffchainWorkerApi<Block>
+		+ sp_block_builder::BlockBuilder<Block>
+		+ cumulus_primitives_core::CollectCollationInfo<Block>
+		+ sp_consensus_aura::AuraApi<Block, <<AuraId as AppCrypto>::Pair as Pair>::Public>
+		+ pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>
+		+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>
+		+ cumulus_primitives_aura::AuraUnincludedSegmentApi<Block>
+	{
+	}
+	impl<
+			T: sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+				+ sp_api::Metadata<Block>
+				+ sp_session::SessionKeys<Block>
+				+ sp_api::ApiExt<Block>
+				+ sp_offchain::OffchainWorkerApi<Block>
+				+ sp_block_builder::BlockBuilder<Block>
+				+ cumulus_primitives_core::CollectCollationInfo<Block>
+				+ sp_consensus_aura::AuraApi<Block, <<AuraId as AppCrypto>::Pair as Pair>::Public>
+				+ pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>
+				+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>
+				+ cumulus_primitives_aura::AuraUnincludedSegmentApi<Block>,
+			RuntimeApi,
+		> RuntimeApiExt<RuntimeApi> for T
+	{
+	}
 }
-impl<
-		T: sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
-			+ sp_api::Metadata<Block>
-			+ sp_session::SessionKeys<Block>
-			+ sp_api::ApiExt<Block>
-			+ sp_offchain::OffchainWorkerApi<Block>
-			+ sp_block_builder::BlockBuilder<Block>
-			+ cumulus_primitives_core::CollectCollationInfo<Block>
-			+ sp_consensus_aura::AuraApi<Block, <<AuraId as AppCrypto>::Pair as Pair>::Public>
-			+ pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>
-			+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>
-			+ cumulus_primitives_aura::AuraUnincludedSegmentApi<Block>,
-		RuntimeApi,
-	> RuntimeApiExt<RuntimeApi> for T
-{
+
+#[cfg(feature = "ismp")]
+mod runtime_api_ext {
```

Comment:
I wonder if instead of repeating the whole trait, its just:

```
IsmpExt<RuntimeApi>: RuntimeApiExt<RuntimeApi> 
  + ismp_parachain_runtime_api::IsmpParachainApi<Block>
  + pallet_ismp_runtime_api::IsmpRuntimeApi<Block, H256>
```

Might still need to have some conditional compilation and type aliases, but might be less code overall.

Reply:
I tried a few variations of this, but unfortunately, I can't get it working.

Reply:
It largely has to do with the fact that trait aliases are experimental. If we built the node with +nightly, it would be much cleaner.

Reply:
Aaah, fair enough, no problem.